nomad/stream/ndjson.go

```diff
@@ -0,0 +1,105 @@
+package stream
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+var (
+	// NDJsonHeartbeat is the NDJson to send as a heartbeat
+	// Avoids creating many heartbeat instances
+	NDJsonHeartbeat = &structs.NDJson{Data: []byte("{}\n")}
+)
+
+// NDJsonStream is used to send new line delimited JSON and heartbeats
+// to a destination (out channel)
+type NDJsonStream struct {
+	// mu locks during send and protects running
+	mu sync.Mutex
+
+	out chan<- *structs.NDJson
+
+	// heartbeat is the interval to send heartbeat messages to keep a connection
+	// open.
+	heartbeat *time.Ticker
+
+	publishCh chan structs.NDJson
+	exitCh    chan struct{}
+
+	running bool
+}
+
+// NewNNewNDJsonStream creates a new NDJson stream that will output NDJson structs
+// to the passed output channel
+func NewNDJsonStream(out chan<- *structs.NDJson, heartbeat time.Duration) *NDJsonStream {
+	return &NDJsonStream{
+		out:       out,
+		heartbeat: time.NewTicker(heartbeat),
+		exitCh:    make(chan struct{}),
+		publishCh: make(chan structs.NDJson),
+	}
+}
+
+// Run starts a long lived goroutine that handles sending
+// heartbeats and processed json objects to the streams out channel as well
+func (n *NDJsonStream) Run(ctx context.Context) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if n.running {
+		return
+	}
+	n.running = true
+
+	go n.run(ctx)
```

Comment:
Have you considered starting the run/heartbeat goroutine immediately in the constructor - it eliminates the need for maintaining running and exitCh (in favor of using ctx directly).

I know that some discourage starting goroutines in a constructor - however, currently this class will be not functioning as expected without the `Run` function - so it seems adhering to the rules just adds boilerplate.

Reply:
I think that approach is much better, will update